src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -222,6 +203,55 @@ public void StopProcessingNextRequest()
             }
         }
 
+        // https://tools.ietf.org/html/rfc7540#section-9.2
+        // Some of these could not be checked in advance. Fail before using the connection.
+        private void ValidateTlsRequirements()
+        {
+            var tlsFeature = ConnectionFeatures?.Get<ITlsHandshakeFeature>();
+            if (tlsFeature == null)
+            {
+                // Not using TLS at all.
+                return;
+            }
+
+            if (tlsFeature.Protocol < SslProtocols.Tls12)
+            {
+                throw new Http2ConnectionErrorException(CoreStrings.FormatHttp2ErrorMinTlsVersion(tlsFeature.Protocol), Http2ErrorCode.INADEQUATE_SECURITY);
+            }
+        }
+
+        private async Task<bool> TryReadPrefaceAsync()
```

Comment:
Are we measuring performance yet? I'm guessing splitting this out into another async method could be very costly.

Reply:
Not yet.

Reply:
Note it's only once per connection.

Reply:
Ah, then that's likely ok.